cni/pkg/repair/netns.go

```diff
@@ -33,46 +31,49 @@ func getPidNamespace(pid int) string {
 	return "/host/proc/" + strconv.Itoa(pid) + "/ns/net"
 }
 
-func runInHost[T any](f func() (T, error)) (T, error) {
+func runInNS[T any](pidNs string, f func() (T, error)) (T, error) {
 	var res T
-	ns, err := netns.GetNS(getPidNamespace(1))
+	ns, err := netns.GetNS(pidNs)
 	if err != nil {
-		return res, fmt.Errorf("failed to get host network: %v", err)
+		return res, fmt.Errorf("failed to get network namespace %v: %v", pidNs, err)
 	}
 	err = ns.Do(func(ns netns.NetNS) error {
 		var err error
 		res, err = f()
 		return err
 	})
 	if err != nil {
-		return res, fmt.Errorf("in host network: %v", err)
+		return res, fmt.Errorf("in network namespace %v: %v", ns, err)
 	}
 	return res, nil
 }
 
-func findNetworkIDByIP(ip string) (int, error) {
-	link, err := getLinkWithDestinationOf(ip)
-	if err != nil {
-		return 0, fmt.Errorf("find link for %v: %v", ip, err)
-	}
-	return link.Attrs().NetNsID, nil
+func runInHost[T any](f func() (T, error)) (T, error) {
+	return runInNS(getPidNamespace(1), f)
 }
 
-func getLinkWithDestinationOf(ip string) (netlink.Link, error) {
-	routes, err := netlink.RouteListFiltered(
-		netlink.FAMILY_V4,
-		&netlink.Route{Dst: &net.IPNet{IP: net.ParseIP(ip), Mask: net.CIDRMask(32, 32)}},
-		netlink.RT_FILTER_DST)
-	if err != nil {
-		return nil, err
+func getInterfaceAddr(interfaceName string) (addr string, err error) {
+	var (
+		ief   *net.Interface
+		addrs []net.Addr
+	)
+	if ief, err = net.InterfaceByName(interfaceName); err != nil {
+		return
 	}
-
-	if len(routes) == 0 {
-		return nil, fmt.Errorf("no routes found for %s", ip)
+	if addrs, err = ief.Addrs(); err != nil {
+		return
 	}
-
-	linkIndex := routes[0].LinkIndex
-	return netlink.LinkByIndex(linkIndex)
+	for _, addr := range addrs {
+		switch v := addr.(type) {
+		case *net.IPNet:
+			if v.IP.To4() != nil {
```

Comment:
This will return first IP, then we compare to the pod.status. Not sure in dual stack those are guaranteed to both put the same version first

Reply:
I'll change the logic to pass in the pod IP and check each interface for a matching ip directly.